File: app/Payload/gladCodeAPI.c
/*
contém as funções que podem ser chamadas diretamente pelo usuário.
Estas funções somente enviam sinais do cliente para o servidor. O servidor é quem executa as funções e devolve a resposta.
Desta maneira somente com a sintaxe dos sinais, pode-se fazer o port facilmente para qualquer linguagem.
*/

float getSimTime(){
    char r[10];
    sendMessage("getSimTime", r);
    return atof(r);
}

void setSTR(int arg){
    char message[50], response[10];
    sprintf(message, "setSTR %i", arg);
    sendMessage(message,response);
}

int getSTR(){
    char response[10];
    sendMessage("getSTR", response);
    return atoi(response);
}

void setAGI(int arg){
    char message[50], response[10];
    sprintf(message, "setAGI %i", arg);
    sendMessage(message,response);
}

int getAGI(){
    char response[10];
    sendMessage("getAGI", response);
    return atoi(response);
}

void setINT(int arg){
    char message[50], response[10];
    sprintf(message, "setINT %i", arg);
    sendMessage(message,response);
}

int getINT(){
    char response[10];
    sendMessage("getINT", response);
    return atoi(response);
}


void setName(char *name){
    char m[100], *r, bla[100];
    int i;
    //faz isso pra trocar todos espeços por #
    strcpy(bla,name);
    for (i=0 ; name[i] != '\0' ; i++){
        if (name[i] == ' ')
            bla[i] = '#';
    }
    sprintf(m, "setName %s", bla);
    sendMessage(m, r);
}

char* getName(){
    char *response = (char*)malloc(sizeof(char)*100);
    sendMessage("getName", response);
    return response;
}

int upgradeSTR(int n){
    char message[50], response[10];
    sprintf(message, "upgradeSTR %i", n);
    sendMessage(message,response);
    return atoi(response);
}

int upgradeAGI(int n){
    char message[50], response[10];
    sprintf(message, "upgradeAGI %i", n);
    sendMessage(message,response);
    return atoi(response);
}

int upgradeINT(int n){
    char message[50], response[10];
    sprintf(message, "upgradeINT %i", n);
    sendMessage(message,response);
    return atoi(response);
}

float stepForward(){
    char r[10];
    sendMessage("stepForward", r);
    return atof(r);
}

float stepBack(){
    char r[10];
    sendMessage("stepBack", r);
    return atof(r);
}

float stepLeft(){
    char r[10];
    sendMessage("stepLeft", r);
    return atof(r);
}

float stepRight(){
    char r[10];
    sendMessage("stepRight", r);
    return atof(r);
}

float turnLeft(float ang){
    char m[100], r[10];
    sprintf(m, "turnLeft %f", ang);
    sendMessage(m, r);
    return atof(r);
}

float turnRight(float ang){
    char m[100], r[10];
    sprintf(m, "turnRight %f", ang);
    sendMessage(m, r);
    return atof(r);
}

void turn(float ang){
    char m[100], r[10];
    sprintf(m, "turn %f", ang);
    sendMessage(m, r);
}

int turnTo(float x, float y){
    char m[100], r[10];
    sprintf(m, "turnTo %f %f", x, y);
    sendMessage(m, r);
    return atoi(r);
}

int turnToTarget(){
    char m[100], r[10];
    sprintf(m, "turnToTarget");
    sendMessage(m, r);
    return atoi(r);
}

int turnToAngle(float ang){
    char m[100], r[10];
    sprintf(m, "turnToAngle %f", ang);
    sendMessage(m, r);
    return atoi(r);
}

void moveForward(float p){
    char m[100], r[10];
    sprintf(m, "moveForward %f", p);
    sendMessage(m, r);
}

int moveTo(float x, float y){
    char m[100], r[10];
    sprintf(m, "moveTo %f %f", x, y);
    sendMessage(m, r);
    return atoi(r);
}

int moveToTarget(){
    char r[10];
    sendMessage("moveToTarget", r);
    return atoi(r);
}

float getX(){
    char r[10];
    sendMessage("getX", r);
    return atof(r);
}

float getY(){
    char r[10];
    sendMessage("getY", r);
    return atof(r);
}

float getHp(){
    char r[10];
    sendMessage("getHp", r);
    return atof(r);
}

float getAp(){
    char r[10];
    sendMessage("getAp", r);
    return atof(r);
}

float getSpeed(){
    char r[10];
    sendMessage("getSpeed", r);
    return atof(r);
}

float getHead(){
    char r[10];
    sendMessage("getHead", r);
    return atof(r);
}

float getDist(float x, float y){
    char m[100], r[10];
    sprintf(m, "getDist %f %f", x, y);
    sendMessage(m, r);
    return atof(r);
}

float getDistToTarget(){
    char r[10];
    sendMessage("getDistToTarget", r);
    return atof(r);
}

float getAngle(float x, float y){
    char m[100], r[10];
    sprintf(m, "getAngle %f %f", x, y);
    sendMessage(m, r);
    return atof(r);
}

int howManyEnemies(){
    char r[10];
    sendMessage("howManyEnemies", r);
    return atoi(r);
}

int getCloseEnemy(){
    char r[10];
    sendMessage("getCloseEnemy", r);
    return atoi(r);
}

int getFarEnemy(){
    char r[10];
    sendMessage("getFarEnemy", r);
    return atoi(r);
}

int getLowHp(){
    char r[10];
    sendMessage("getLowHp", r);
    return atoi(r);
}

int getHighHp(){
    char r[10];
    sendMessage("getHighHp", r);
    return atoi(r);
}

float getTargetX(){
    char r[10];
    sendMessage("getTargetX", r);
    return atof(r);
}

float getTargetY(){
    char r[10];
    sendMessage("getTargetY", r);
    return atof(r);
}

float getTargetHealth(){
    char r[10];
    sendMessage("getTargetHealth", r);
    return atof(r);
}

float getTargetSpeed(){
    char r[10];
    sendMessage("getTargetSpeed", r);
    return atof(r);
}

float getTargetHead(){
    char r[10];
    sendMessage("getTargetHead", r);
    return atof(r);
}

int doYouSeeMe(){
    char r[10];
    sendMessage("doYouSeeMe", r);
    return atoi(r);
}

int isTargetVisible(){
    char r[10];
    sendMessage("isTargetVisible", r);
    return atoi(r);
}

void attackMelee(){
    char r[10];
    sendMessage("attackMelee", r);
}

int attackRanged(float x, float y){
    char m[100], r[10];
    sprintf(m, "attackRanged %f %f", x, y);
    sendMessage(m, r);
    return atoi(r);
}

float getLastHitTime(){
    char response[10];
    sendMessage("getLastHitTime", response);
    return atof(response);
}

float getLastHitAngle(){
    char response[10];
    sendMessage("getLastHitAngle", response);
    return atof(response);
}

int turnToLastHit(){
    char r[10];
    sendMessage("turnToLastHit", r);
    return atoi(r);
}

int getHit(){
    char r[10];
    sendMessage("getHit", r);
    return atoi(r);
}

float getSafeRadius(){
    char response[10];
    sendMessage("getSafeRadius", response);
    return atof(response);
}

int isSafeHere(){
    char r[10];
    sendMessage("isSafeHere", r);
    return atoi(r);
}

int isSafeThere(float x, float y){
    char m[100], r[10];
    sprintf(m, "isSafeThere %f %f", x, y);
    sendMessage(m, r);
    return atoi(r);
}

int fireball(float x, float y){
    char m[100], r[10];
    sprintf(m, "fireball %f %f", x, y);
    sendMessage(m, r);
    return atoi(r);
}

int teleport(float x, float y){
    char m[100], r[10];
    sprintf(m, "teleport %f %f", x, y);
    sendMessage(m, r);
    return atoi(r);
}

int block(){
    char r[10];
    sendMessage("block", r);
    return atoi(r);
}

int ambush(){
    char r[10];
    sendMessage("ambush", r);
    return atoi(r);
}

int assassinate(float x, float y){
    char m[100], r[10];
    sprintf(m, "assassinate %f %f", x, y);
    sendMessage(m, r);
    return atoi(r);
}

int charge(){
    char r[10];
    sendMessage("charge", r);
    return atoi(r);
}

float getBlockTimeLeft(){
    char r[10];
    sendMessage("getBlockTimeLeft", r);
    return atof(r);
}

float getAmbushTimeLeft(){
    char r[10];
    sendMessage("getAmbushTimeLeft", r);
    return atof(r);
}

float getBurnTimeLeft(){
    char r[10];
    sendMessage("getBurnTimeLeft", r);
    return atof(r);
}

void setSpritesheet(char *str){
    //faz nada mesmo
}

int isStunned(){
    char r[10];
    sendMessage("isStunned", r);
    return atoi(r);
}

int isBurning(){
    char r[10];
    sendMessage("isBurning", r);
    return atoi(r);
}

int isProtected(){
    char r[10];
    sendMessage("isProtected", r);
    return atoi(r);
}

int isRunning(){
    char r[10];
    sendMessage("isRunning", r);
    return atoi(r);
}

int isSlowed(){
    char r[10];
    sendMessage("isSlowed", r);
    return atoi(r);
}
void speak(char *format, ...){
    va_list arg;
    va_start(arg, format);

    char message[1000] = "";
    while (*format != '\0' && strlen(message) < 1000) {
        if (*format == '%') {
            format++;
            if (*format == '%')
                sprintf(message, "%s%%", message);
            else if (*format == 'c')
                sprintf(message, "%s%c", message, va_arg(arg, int));
            else if (*format == 's')
                sprintf(message, "%s%s", message, va_arg(arg, char*));
            else if (*format == 'i' || *format == 'd')
                sprintf(message, "%s%i", message, va_arg(arg, int));
            else if (*format == 'f')
                sprintf(message, "%s%f", message, va_arg(arg, double));
            else if (*format == '.'){
                char f[10] = "%s%.0f";
                f[4] = *(format + 1);
                format += 2;
                sprintf(message, f, message, va_arg(arg, double));
            }
        }
        else {
            sprintf(message, "%s%c", message, *format);
        }
        format++;
    }
    va_end(arg);
    message[999] = '\0';

    char m[1000], apiMessage[1000], r[10];
    strncpy(m, message, 1000);
    sprintf(apiMessage, "speak %s", m);
    apiMessage[999] = '\0';
    sendMessage(apiMessage, r);
}

int getLvl(){
    char r[10];
    sendMessage("getLvl", r);
    return atoi(r);
}

void breakpoint(char *message){
    char m[256], r[10];
    sprintf(m, "breakpoint %s", message);
    sendMessage(m, r);
}

void setPosition(float x, float y){
    char message[50], response[10];
    sprintf(message, "setPosition %f %f", x, y);
    sendMessage(message,response);
}

void setHp(float hp){
    char message[50], response[10];
    sprintf(message, "setHp %f", hp);
    sendMessage(message,response);
}

void setAp(float ap){
    char message[50], response[10];
    sprintf(message, "setAp %f", ap);
    sendMessage(message,response);
}

void lvlUp(int n){
    char message[50], response[10];
    sprintf(message, "lvlUp %i", n);
    sendMessage(message,response);
}

int useItem(char *str){
    char m[100], r[10];
    sprintf(m, "useItem %s", str);
    sendMessage(m, r);
    return atoi(r);
}

void setSlots(char *str){
    char message[50], response[10];
    sprintf(message, "setSlots %s", str);
    sendMessage(message,response);
}

int isItemReady(char *str){
    char m[100], r[10];
    sprintf(m, "isItemReady %s", str);
    sendMessage(m, r);
    return atoi(r);
}

File: app/Payload/gladCodeAPI.py
'''
contém as funções que podem ser chamadas diretamente pelo usuário.
Estas funções somente enviam sinais do cliente para o servidor. O servidor é quem executa as funções e devolve a resposta.
Desta maneira somente com a sintaxe dos sinais, pode-se fazer o port facilmente para qualquer linguagem.
'''

from gladCodeCore import *

def getSimTime():
    return float(sendMessage("getSimTime"))

def setSTR(arg):
    sendMessage("setSTR {}".format(arg))

def getSTR():
    return int(sendMessage("getSTR"))

def setAGI(arg):
    sendMessage("setAGI {}".format(arg))

def getAGI():
    return int(sendMessage("getAGI"))

def setINT(arg):
    sendMessage("setINT {}".format(arg))

def getINT():
    return int(sendMessage("getINT"))

def setName(name):
    name = name.replace(" ", "#")
    sendMessage("setName {}".format(name))

def getName():
    return sendMessage("getName")

def upgradeSTR(n):
    return bool(int(sendMessage("upgradeSTR {}".format(n))))

def upgradeAGI(n):
    return bool(int(sendMessage("upgradeAGI {}".format(n))))

def upgradeINT(n):
    return bool(int(sendMessage("upgradeINT {}".format(n))))

def stepForward():
    return float(sendMessage("stepForward"))

def stepBack():
    return float(sendMessage("stepBack"))

def stepLeft():
    return float(sendMessage("stepLeft"))

def stepRight():
    return float(sendMessage("stepRight"))

def turnLeft(ang):
    return float(sendMessage("turnLeft {}".format(ang)))

def turnRight(ang):
    return float(sendMessage("turnRight {}".format(ang)))

def turn(ang):
    sendMessage("turn {}".format(ang))

def turnTo(x, y):
    return bool(int(sendMessage("turnTo {} {}".format(x,y))))

def turnToTarget():
    return bool(int(sendMessage("turnToTarget")))

def turnToAngle(ang):
    return bool(int(sendMessage("turnToAngle {}".format(ang))))

def moveForward(p):
    sendMessage("moveForward {}".format(p))

def moveTo(x, y):
    return bool(int(sendMessage("moveTo {} {}".format(x,y))))

def moveToTarget():
    return bool(int(sendMessage("moveToTarget")))

def getX():
    return float(sendMessage("getX"))

def getY():
    return float(sendMessage("getY"))

def getHp():
    return float(sendMessage("getHp"))

def getAp():
    return float(sendMessage("getAp"))

def getSpeed():
    return float(sendMessage("getSpeed"))

def getHead():
    return float(sendMessage("getHead"))

def getDist(x, y):
    return float(sendMessage("getDist {} {}".format(x, y)))

def getDistToTarget():
    resp = float(sendMessage("getDistToTarget"))
    if resp < 0:
        return False
    else:
        return resp

def getAngle(x, y):
    return float(sendMessage("getAngle {} {}".format(x, y)))


def howManyEnemies():
    return int(sendMessage("howManyEnemies"))

def getCloseEnemy():
    return bool(int(sendMessage("getCloseEnemy")))

def getFarEnemy():
    return bool(int(sendMessage("getFarEnemy")))

def getLowHp():
    return bool(int(sendMessage("getLowHp")))

def getHighHp():
    return bool(int(sendMessage("getHighHp")))

def getTargetX():
    resp = float(sendMessage("getTargetX"))
    if resp < 0:
        return False
    else:
        return resp

def getTargetY():
    resp = float(sendMessage("getTargetY"))
    if resp < 0:
        return False
    else:
        return resp

def getTargetHealth():
    return float(sendMessage("getTargetHealth"))

def getTargetSpeed():
    resp = float(sendMessage("getTargetSpeed"))
    if resp < 0:
        return False
    else:
        return resp

def getTargetHead():
    resp = float(sendMessage("getTargetHead"))
    if resp < 0:
        return False
    else:
        return resp

def doYouSeeMe():
    return bool(int(sendMessage("doYouSeeMe")))

def isTargetVisible():
    return bool(int(sendMessage("isTargetVisible")))

def attackMelee():
    sendMessage("attackMelee")

def attackRanged(x, y):
    return bool(int(sendMessage("attackRanged {} {}".format(x, y))))

def getLastHitTime():
    return float(sendMessage("getLastHitTime"))

def getLastHitAngle():
    return float(sendMessage("getLastHitAngle"))

def turnToLastHit():
    return bool(int(sendMessage("turnToLastHit")))

def getHit():
    return bool(int(sendMessage("getHit")))

def getSafeRadius():
    return float(sendMessage("getSafeRadius"))

def isSafeHere():
    return bool(int(sendMessage("isSafeHere")))

def isSafeThere(x, y):
    return bool(int(sendMessage("isSafeThere {} {}".format(x, y))))

def fireball(x, y):
    return bool(int(sendMessage("fireball {} {}".format(x, y))))

def teleport(x, y):
    return bool(int(sendMessage("teleport {} {}".format(x, y))))

def block():
    return bool(int(sendMessage("block")))

def ambush():
    return bool(int(sendMessage("ambush")))

def assassinate(x, y):
    return bool(int(sendMessage("assassinate {} {}".format(x, y))))

def charge():
    return bool(int(sendMessage("charge")))

def getBlockTimeLeft():
    return float(sendMessage("getBlockTimeLeft"))

def getAmbushTimeLeft():
    return float(sendMessage("getAmbushTimeLeft"))

def getBurnTimeLeft():
    return float(sendMessage("getBurnTimeLeft"))

def setSpritesheet(str):
    #faz nada mesmo
    str = str

def isStunned():
    return bool(int(sendMessage("isStunned")))

def isBurning():
    return bool(int(sendMessage("isBurning")))

def isProtected():
    return bool(int(sendMessage("isProtected")))

def isRunning():
    return bool(int(sendMessage("isRunning")))

def isSlowed():
    return bool(int(sendMessage("isSlowed")))

def speak(message):
    sendMessage("speak {}".format(message))

def getLvl():
    return int(sendMessage("getLvl"))

def breakpoint(message):
    sendMessage("breakpoint {}".format(message))

def setPosition(x, y):
    sendMessage("setPosition {} {}".format(x, y))

def setHp(hp):
    sendMessage("setHp {}".format(hp))

def setAp(ap):
    sendMessage("setAp {}".format(ap))

def lvlUp(n):
    sendMessage("lvlUp {}".format(n))

def useItem(item):
    return bool(int(sendMessage("useItem {}".format(item))))

def setSlots(slots):
    sendMessage("setSlots {}".format(slots))

def isItemReady(item):
    return bool(int(sendMessage("isItemReady {}".format(item))))
